business/openid_auth.go

```diff
@@ -459,6 +548,134 @@ func ValidateOpenIdState(openIdParams *OpenIdCallbackParams) (validationFailure
+		return fmt.Errorf("the OpenId token has unexpected issuer claim; got iss = '%s'", issuerClaim)
+	}
+
+	// Check the aud claim contains our client-id
+	checkAzpClaim := false
+	if audienceClaim, ok := idTokenClaims["aud"]; !ok {
+		return errors.New("the OpenId token has no aud claim")
+	} else {
+		switch ac := audienceClaim.(type) {
+		case string:
+			if oidCfg.ClientId != ac {
+				return fmt.Errorf("the OpenId token is not targeted for Kiali; got aud = '%s'", audienceClaim)
+			}
+		case []string:
+			audFound := false
+			for _, audItem := range ac {
+				if oidCfg.ClientId == audItem {
+					audFound = true
+				}
+			}
+			if !audFound {
+				return fmt.Errorf("the OpenId token is not targeted for Kiali; got aud = %v", audienceClaim)
+			}
+
+			// The OIDC Spec says that if the aud claim contains multiple audiences, we "SHOULD" check
+			// the azp claim is present. In Kiali, there is currently no known reason to omit this
+			// check, so we do it.
+			checkAzpClaim = true
+		default:
+			return fmt.Errorf("the OpenId token has an unexpected audience claim; got '%v'", audienceClaim)
+		}
+	}
+
+	if checkAzpClaim {
+		// Check that the azp claim is present and contains our client_id
+		if authorizedPartyClaim, ok := idTokenClaims["azp"].(string); !ok {
+			return fmt.Errorf("the OpenId token has an invalid 'azp' claim")
+		} else if oidCfg.ClientId != authorizedPartyClaim {
+			return fmt.Errorf("the OpenId token is not targeted for Kiali; got azp = %v", authorizedPartyClaim)
+		}
+	}
+
+	// Currently, we only support tokens with an RSA signature with SHA-256, which is the default in the OIDC spec
+	if algHeader, ok := openIdParams.ParsedIdToken.Header["alg"].(string); !ok || algHeader != "RS256" {
+		return fmt.Errorf("the OpenId token has unexpected alg header claim; got alg = '%s'", algHeader)
+	}
+
+	// Check iat (issued at) claim
+	if iatClaim, ok := idTokenClaims["iat"]; !ok {
+		return errors.New("the OpenId token has no iat claim or is invalid")
+	} else {
+		parsedIat, parseErr := parseTimeClaim(iatClaim)
+		if parseErr != nil {
+			return fmt.Errorf("the OpenId token has an invalid iat claim: %w", parseErr)
+		}
+		if parsedIat == 0 {
+			// This is weird. This would mean an invalid type
+			return fmt.Errorf("the OpenId token has an invalid value in the iat claim; got '%v'", iatClaim)
+		}
+
+		// Let's do the minimal check to ensure that the token wasn't issued in the future
+		// we add a little offset to "now" to add one minute tolerance
+		iatTime := time.Unix(parsedIat, 0)
+		nowTime := util.Clock.Now().Add(60 * time.Second)
+		if iatTime.After(nowTime) {
+			return fmt.Errorf("we don't like people living in the future - enjoy the present!; iat = '%d'", parsedIat)
+		}
+	}
+
+	// Check exp (expiration time) claim
+	// The OIDC spec says: "The current time MUST be before the time represented by the exp Claim"
+	// No tolerance for this check.
+	if !util.Clock.Now().Before(openIdParams.ExpiresOn) {
```

Comment:
I think `if util.Clock.Now().After(openIdParams.ExpiresOn)` might look better here, wdyt?

Reply:
I also though about it, but I would need something like `AfterOrEquals`. I'm not sure if there is that function. Otherwise, I could try `util.Clock.Now().After(openIdParams.ExpiresOn) || util.Clock.Now().Equal(openIdParams.ExpiresOn)`

Thoughts?

Reply:
You are correct, no need to change then :)